distributed/worker.py

```diff
@@ -2182,73 +2182,86 @@ def execute(self, key, report=False):
-
-        # Pause worker threads if above 80% memory use
         if self.memory_pause_fraction and frac > self.memory_pause_fraction:
-            # Try to free some memory while in paused state
-            self._throttled_gc.collect()
+            # We intentionally restrain ourselves from calling
+            # self._throttled_gc.collect() here as we would rather call it just
+            # after self.data.fast.evict() to give it a higher likelyhood to
+            # actually free memory.
             if not self.paused:
                 logger.warn("Worker is at %d%% memory usage. Pausing worker.  "
                             "Process memory: %s -- Worker memory limit: %s",
                             int(frac * 100),
-                            format_bytes(proc.memory_info().rss),
+                            format_bytes(memory),
                             format_bytes(self.memory_limit))
                 self.paused = True
         elif self.paused:
             logger.warn("Worker is at %d%% memory usage. Resuming worker. "
                         "Process memory: %s -- Worker memory limit: %s",
                         int(frac * 100),
-                        format_bytes(proc.memory_info().rss),
+                        format_bytes(memory),
                         format_bytes(self.memory_limit))
             self.paused = False
             self.ensure_computing()
 
+    @gen.coroutine
+    def memory_monitor(self):
+        """ Track this process's memory usage and act accordingly
+
+        If we rise above 70% memory use, dump data to disk.
+
+        If we rise above 80% memory use, pause execution of new tasks
+        """
+        total = 0
+        if self._memory_monitoring:
+            raise gen.Return(total)
+        self._memory_monitoring = True
+
+        proc = psutil.Process()
+        memory = proc.memory_info().rss
+        frac = memory / self.memory_limit
+
+        # Pause worker threads if above 80% memory use
+        self._check_pause_resume(memory)
+
         # Dump data to disk if above 70%
         if self.memory_spill_fraction and frac > self.memory_spill_fraction:
             target = self.memory_limit * self.memory_target_fraction
             count = 0
-            need = memory - target
             while memory > target:
                 if not self.data.fast:
-                    logger.warn("Memory use is high but worker has no data "
-                                "to store to disk.  Perhaps some other process "
-                                "is leaking memory?  Process memory: %s -- "
-                                "Worker memory limit: %s",
-                                format_bytes(proc.memory_info().rss),
-                                format_bytes(self.memory_limit))
+                    self._throttled_gc.collect()
+                    yield gen.moment
+                    memory = proc.memory_info().rss
+                    self._check_pause_resume(memory)
```

Comment:
Why do we repeat these checks?  A (not necessarily better) alternative would be to wait for the next iteration of the `memory_monitor`.